src/cryptonote_core/loki_name_system.cpp

```diff
@@ -1483,60 +1489,92 @@ CREATE TABLE IF NOT EXISTS "settings" (
     "id" INTEGER PRIMARY KEY NOT NULL,
     "top_height" INTEGER NOT NULL,
     "top_hash" VARCHAR NOT NULL,
-    "version" INTEGER NOT NULL
+    "version" INTEGER NOT NULL,
+    "pruned_height" INTEGER NOT NULL DEFAULT 0
 );
 
-CREATE TABLE IF NOT EXISTS "mappings" (
-    "id" INTEGER PRIMARY KEY NOT NULL,
-    "type" INTEGER NOT NULL,
-    "name_hash" VARCHAR NOT NULL,
-    "encrypted_value" BLOB NOT NULL,
-    "txid" BLOB NOT NULL,
-    "prev_txid" BLOB NOT NULL,
-    "register_height" INTEGER NOT NULL,
-    "owner_id" INTEGER NOT NULL REFERENCES "owner" ("id"),
-    "backup_owner_id" INTEGER REFERENCES "owner" ("id"),
-    "update_height" INTEGER NOT NULL DEFAULT "register_height",
-    "expiration_height" INTEGER
-);
-CREATE UNIQUE INDEX IF NOT EXISTS "name_hash_type_id" ON mappings("name_hash", "type");
+CREATE TABLE IF NOT EXISTS "mappings" ()" + mappings_columns + R"();
 CREATE INDEX IF NOT EXISTS "owner_id_index" ON mappings("owner_id");
 CREATE INDEX IF NOT EXISTS "backup_owner_id_index" ON mappings("backup_owner_index");
 )";
 
   char *table_err_msg = nullptr;
-  int table_created   = sqlite3_exec(db, BUILD_TABLE_SQL, nullptr /*callback*/, nullptr /*callback context*/, &table_err_msg);
+  int table_created   = sqlite3_exec(lns_db.db, BUILD_TABLE_SQL.c_str(), nullptr /*callback*/, nullptr /*callback context*/, &table_err_msg);
   if (table_created != SQLITE_OK)
   {
     MERROR("Can not generate SQL table for LNS: " << (table_err_msg ? table_err_msg : "??"));
     sqlite3_free(table_err_msg);
     return false;
   }
 
+  // In Loki 8 we dropped some columns that are no longer needed, but SQLite can't do this easily:
+  // instead we have to manually recreate the table, so check it and see if the prev_txid or
+  // register_height columns still exist: if so, we need to recreate.
+  bool need_mappings_migration = false;
```

Comment:
There are 2 places we do migration and its sort of weird that we want to build default tables and then check migration again afterwards.

Feels like what we want is to load the tables, check version then update, avoiding building a default table unless we know we have to (lns.db itself doesn't exist yet).

Reply:
> There are 2 places we do migration and its sort of weird that we want to build default tables and then check migration again afterwards.

The reason for doing it this way is that we're doing a `CREATE TABLE IF NOT EXISTS` which means the query will go ahead just fine it already exist with the old structure, but if it *doesn't* then we create it.

The `PRAGMA table_info` should be an extremely fast query, and only done once during initialization to check whether the table we have is the old one (with prev_txid/register_height) or not.

I did not like this messiness: `ALTER TABLE "mappings" DROP COLUMN "prev_txid"` is how I tried to do it at first, only to discover that sqlite3 does not support dropping columns from a table at all (the create-a-new-table-and-copy-it is the documented solution to do this).

Reply:
Oh, re: the settings version: I see that as being for data value migration (e.g. where we needed to go compute the update heights by scanning the database), whereas this chunk of code is only for database schema changes.  It just happens that sqlite3 requires a mess to drop a column. :(

Reply:
Yeah, in the long run it doesn't matter too much, essentially can drop the upgrade code over time as everyone upgrades.